docs/rfcs/20220119_raft_replication_simplification.md

```diff
@@ -0,0 +1,162 @@
+also reduce the disk usage (in some cluster with limited memory cache hit 
+ratio is not always 1.0 even if there are no reads other than reads issued by 
+`replicate_stm`).
+
+### Backpressure
+
+In current `quorum_ack` replication implementation the backpressure is 
+propagated from the followers to the `raft::replicate` caller 
+(Kafka Produce request handler effectively). Currently the backpressure 
+propagation is based on the per follower semaphore since RPC layer 
+lacks backpressure propagation mechanism. 
+When `replicate_entries` stm is unable to acquire follower dispatch units 
+it will wait holding `replicate_batcher` memory reservation lock so that 
+it can not accept writes indefinitely. When `replicate_batcher` memory 
+reservation semaphore is exhausted it prevents first stage of `replicate_stages` 
+from finishing not allowing for more requests to be processed. 
+
+Current `leader_ack` mechanism lacks backpressure propagation. 
+Leader can be arbitrary large number of entries ahead of followers. 
+This solution may be a problem since user has an impression that data are 
+replicated while the data are actually stored only on the leader. 
+
+In the proposed solution we are going to handle back pressure in the same way 
+for both consistency level. This way there will be a guarantee that followers 
+can always keep up with the leader.
+
+
+### Memory usage
+
+Memory usage control will be achieved in the same way as it is done right now 
+i.e. we will not finish first stage of replication when `replicate_batcher` 
+memory semaphore is exhausted. This way we will prevent Kafka layer from 
+buffering indefinite number of batches in pending replicate requests. Replicate 
+batcher memory units are released after the request is handed of to the 
+socket so that all the user space buffer can be freed. 
+In future we may consider per shard semaphore controlling the total amount 
+of memory used by all replicate batchers.
+
+### Recovery STM coordination - finishing recovery
+
+`recovery_stm` is started every time the leader identifies the follower as 
+being behind. Since recovery stm is currently used for both recovering the 
+follower (when it is actually behind) and delivering follower data replicated 
+with `leader_ack` consistency the `recovery_stm` implementation contains some 
+optimization to minimize the `leader_ack` latency. 
+
+There are two main optimizations:
+- not stopping recovery when last replicate request consistency level is `leader_ack`
+- triggering next dispatch loop immediately after leader disk append
+
+Those optimizations make handling of recovering follower more complex in 
+other scenarios f.e. `leadership_transfer`. 
+
+Since there is no coordination between the two paths that the follower append 
+entries requests are sent from (i.e. `replicate_stm` and `recovery_stm`) those 
+request may be reordered and cause log truncations and redeliveries which lead 
+to increased latency.
```

Comment:
```suggestion
Since there is no coordination between the two paths that the follower append 
entry requests are sent from (i.e. `replicate_stm` and `recovery_stm`), those 
requests may be reordered and cause log truncations and redeliveries, which lead 
to increased latency.
```